File: gnome-extension/poketokenbar@huhgeon.github.io/lib/pet.js
/* pet.js — the companion, living on the desktop.
 *
 * This is the one feature a normal Linux app cannot do on Wayland: a client
 * has no way to place its own window at an absolute position, so the pet could
 * never be dragged or restored where it was left. An extension is not a client
 * — it runs inside the compositor and puts an actor straight on the stage — so
 * the whole limitation is simply absent here.
 *
 * The actor is added to the layout manager's chrome, so it floats above
 * windows the way the macOS panel does, and steps out of the way of anything
 * fullscreen.
 */

import Clutter from 'gi://Clutter';
import GObject from 'gi://GObject';
import St from 'gi://St';

import * as Main from 'resource:///org/gnome/shell/ui/main.js';

import {DEFAULT_QUALITY} from './framecap.js';
import {Sprite} from './sprite.js';

// How far a press may travel and still count as a click rather than a drag.
// Without this every click ends as a one-pixel drag and the popup never opens.
const CLICK_SLOP = 4;

// How far above the pet its caption sits.
const TOOLTIP_GAP = 24;

export const DesktopPet = GObject.registerClass(
class DesktopPet extends St.Widget {
    _init(reader, {onActivate, onSavePosition} = {}) {
        super._init({
            style_class: 'poketokenbar-pet',
            reactive: true,
            track_hover: true,
            can_focus: false,
        });
        this._reader = reader;
        this._onActivate = onActivate ?? (() => {});
        this._onSavePosition = onSavePosition ?? (() => {});

        this._sprite = new Sprite({size: 96});
        this.add_child(this._sprite);

        this._tooltip = new St.Label({style_class: 'poketokenbar-pet-tooltip'});
        this._tooltip.hide();
        // Chrome claims an actor's rectangle for input by default, and a label
        // that takes input right above the pet is a rectangle the pointer keeps
        // falling into: hover showed the tooltip, the tooltip took the pointer,
        // the pet saw a leave, the tooltip went away — a flicker for as long as
        // the mouse sat there. It is a caption; it has no business taking
        // clicks.
        Main.layoutManager.addTopChrome(this._tooltip, {affectsInputRegion: false});

        this._dragging = false;
        this._pressPoint = null;
        this._grab = null;

        this.connect('button-press-event', (_a, event) => this._onPress(event));
        this.connect('motion-event', (_a, event) => this._onMotion(event));
        this.connect('button-release-event', (_a, event) => this._onRelease(event));
        this.connect('enter-event', () => this._showTooltip());
        this.connect('leave-event', () => this._tooltip.hide());
        this.connect('destroy', () => {
            // A grab outlives the actor that took it. Dropped mid-drag — by a
            // disable, or by the pet being switched off — it would leave the
            // session routing every pointer event at an actor that is gone.
            this._releaseGrab();
            this._tooltip.destroy();
        });
    }

    setSize(size) {
        this._sprite.setSize(size);
        this.set_size(size, size);
    }

    setPaused(paused) {
        this._sprite.setPaused(paused);
    }

    update(state) {
        // The pet follows the panel, so a pinned species shows here too.
        this._sprite.setPath(state?.panel?.sprite_path || null);
        // Both always-visible surfaces share one setting, as they do upstream.
        this._sprite.setQuality(state?.config?.animation_quality ?? DEFAULT_QUALITY);
        this._tooltipText = state?.today?.tokens_grouped ?? '';
        if (this._tooltip.visible)
            this._showTooltip();
    }

    /** Place the pet, clamped so it can never be dragged off every monitor.
     *
     * A position saved on a monitor that is no longer attached would otherwise
     * leave the pet invisible with no way to get it back.
     */
    place(x, y) {
        const work = Main.layoutManager.getWorkAreaForMonitor(
            Main.layoutManager.primaryIndex);
        const size = this.get_width() || 96;
        const clampedX = Math.max(work.x, Math.min(x, work.x + work.width - size));
        const clampedY = Math.max(work.y, Math.min(y, work.y + work.height - size));
        this.set_position(Math.round(clampedX), Math.round(clampedY));
    }

    _showTooltip() {
        if (!this._tooltipText)
            return;
        this._tooltip.text = this._tooltipText;
        const [x, y] = this.get_transformed_position();
        const work = Main.layoutManager.getWorkAreaForMonitor(
            Main.layoutManager.primaryIndex);
        // Above the pet, unless there is no room above it — a pet parked under
        // the top bar would otherwise caption itself off the top of the screen.
        const above = y - TOOLTIP_GAP;
        const top = above >= work.y ? above : y + this.get_height();
        this._tooltip.set_position(Math.round(x), Math.round(top));
        this._tooltip.show();
    }

    _onPress(event) {
        const [x, y] = event.get_coords();
        if (event.get_button() === Clutter.BUTTON_SECONDARY) {
            this._onActivate('menu');
            return Clutter.EVENT_STOP;
        }
        this._pressPoint = {x, y, actor: this.get_position()};
        this._dragging = false;
        // Without a grab an actor only hears about the pointer while the
        // pointer is over it. A 96px pet is smaller than the distance a hand
        // moves between two frames, so a drag broke up the moment the cursor
        // outran it — and the button release landed on whatever was underneath
        // instead, leaving the press live. The next time the mouse so much as
        // crossed the pet, it picked the drag straight back up.
        this._takeGrab();
        return Clutter.EVENT_STOP;
    }

    _takeGrab() {
        this._releaseGrab();
        try {
            this._grab = global.stage.grab(this);
        } catch (error) {
            // A refused grab costs a smooth drag, not the pet: without one the
            // press still works, it just stops tracking at the actor's edge.
            logError(error, 'PokeTokenBar: could not grab the pointer');
            this._grab = null;
        }
    }

    _releaseGrab() {
        if (!this._grab)
            return;
        this._grab.dismiss();
        this._grab = null;
    }

    _onMotion(event) {
        if (!this._pressPoint)
            return Clutter.EVENT_PROPAGATE;
        const [x, y] = event.get_coords();
        const dx = x - this._pressPoint.x;
        const dy = y - this._pressPoint.y;
        if (!this._dragging && Math.hypot(dx, dy) < CLICK_SLOP)
            return Clutter.EVENT_STOP;
        this._dragging = true;
        this._tooltip.hide();
        this.place(this._pressPoint.actor[0] + dx, this._pressPoint.actor[1] + dy);
        return Clutter.EVENT_STOP;
    }

    _onRelease(_event) {
        this._releaseGrab();
        if (!this._pressPoint)
            return Clutter.EVENT_PROPAGATE;
        const wasDragging = this._dragging;
        this._pressPoint = null;
        this._dragging = false;
        if (wasDragging) {
            const [x, y] = this.get_position();
            this._onSavePosition(x, y);
        } else {
            this._onActivate('click');
        }
        return Clutter.EVENT_STOP;
    }
});

File: poketokenbar/ui/pet.py
"""The companion, living on the desktop.

A frameless always-on-top window rather than an actor on a compositor's stage,
which is what the GNOME front end gets to use. Windows has no equivalent of that
and no equivalent of Wayland's refusal to let a client place its own window
either, so the plain approach works: a borderless translucent widget positioned
wherever it was last left.

Hover shows today's usage, click opens the main window, drag moves it, and the
position is written back through the daemon's own config so it survives a
reboot.
"""

from __future__ import annotations

from PySide6.QtCore import QPoint, Qt
from PySide6.QtWidgets import QApplication, QVBoxLayout, QWidget

from .widgets import Sprite, label

# How far a press may travel and still count as a click rather than a drag.
# Without it every click ends as a one-pixel drag and the window never opens.
CLICK_SLOP = 4

DEFAULT_SIZE = 96

# Shown until a sprite has been fetched even once.
FALLBACK_GLYPH = "\N{EGG}"


class DesktopPet(QWidget):
    def __init__(self, on_activate=None, on_moved=None) -> None:
        super().__init__()
        self._on_activate = on_activate or (lambda: None)
        self._on_moved = on_moved or (lambda x, y: None)
        self._press: QPoint | None = None
        self._dragging = False
        self._tooltip_text = ""
        # The last path that actually resolved to a file. A sprite the daemon
        # could not download comes through as "", and drawing that meant an
        # invisible window — so the pet vanished and came back as the network
        # flapped, which is the same thing as "it keeps disappearing".
        self._last_sprite = ""

        self.setWindowFlags(
            Qt.FramelessWindowHint
            | Qt.WindowStaysOnTopHint
            # Keeps it off the taskbar and out of Alt-Tab: it is an ornament,
            # not a window someone switches to.
            | Qt.Tool
            # And it must never become the active window. Windows owns a tool
            # window to whichever window of the application was last active, so
            # a pet that takes activation on click becomes owned by the popup —
            # and vanishes with it the moment the popup is closed again. That
            # is the "it disappears when I click it" report.
            | Qt.WindowDoesNotAcceptFocus
        )
        self.setAttribute(Qt.WA_TranslucentBackground)
        # Same reason, for the show() that follows: showing a window normally
        # activates it.
        self.setAttribute(Qt.WA_ShowWithoutActivating)
        self.setFocusPolicy(Qt.NoFocus)

        layout = QVBoxLayout(self)
        layout.setContentsMargins(0, 0, 0, 0)
        self.sprite = Sprite(DEFAULT_SIZE)
        layout.addWidget(self.sprite)
        self.set_size(DEFAULT_SIZE)

    # --- state -------------------------------------------------------------

    def set_size(self, size: int) -> None:
        """Resize the window and the sprite in it.

        Poking the sprite's private size did the first half only: the picture
        had already been scaled when it was loaded, and nothing rescaled it, so
        the slider moved the box and left the Pokemon the size it was.
        """
        self.sprite.set_size(size)
        self.setFixedSize(size, size)

    def update_state(self, state: dict | None) -> None:
        panel = (state or {}).get("panel") or {}
        # Follows the panel, so a pinned species shows here too — the daemon has
        # already resolved which one that is. An empty path is a fetch that has
        # not succeeded yet, never a decision to show nothing, so the last one
        # that worked stays up rather than the pet blanking.
        path = panel.get("sprite_path") or ""
        if path:
            self._last_sprite = path
        # Before the first sprite ever arrives — a fresh install with no
        # network — the glyph is what stands in. An empty translucent window
        # is indistinguishable from the pet being gone.
        self.sprite.set_fallback(FALLBACK_GLYPH)
        self.sprite.set_path(self._last_sprite or None)

        config = (state or {}).get("config") or {}
        size = int(config.get("floating_pet_size") or DEFAULT_SIZE)
        if size != self.width():
            self.set_size(size)
        # Both always-visible surfaces share one quality setting, as upstream's
        # do: a frame costs a redraw wherever it is drawn.
        from .widgets import quality_of

        self.sprite.set_quality(quality_of(config))

        today = (state or {}).get("today") or {}
        self._tooltip_text = today.get("tokens_grouped", "")
        self.setToolTip(self._tooltip_text)

    def place(self, x: int, y: int) -> None:
        """Move the pet, keeping it reachable on some screen.

        Clamped to the screen the position lands on, not to the one the pet is
        currently on — that made the edge of the current monitor a wall, so a
        pet on the primary display could never be dragged onto the second.

        A position on a monitor that is no longer attached still has to end up
        somewhere visible, so a point that belongs to no screen is pulled into
        the nearest one rather than left where it was.
        """
        x, y = int(x), int(y)
        screen = self._screen_for(x, y)
        if screen is None:
            self.move(x, y)
            return
        area = screen.availableGeometry()
        self.move(
            max(area.left(), min(x, area.right() - self.width() + 1)),
            max(area.top(), min(y, area.bottom() - self.height() + 1)),
        )

    def _screen_for(self, x: int, y: int):
        """The screen a position belongs to, else the nearest one.

        Nearest by the distance to each screen's rectangle, so dragging past
        the edge of one monitor hands the pet to whichever is actually next to
        it — including one above or below, which comparing centres gets wrong
        on a stacked arrangement.
        """
        application = QApplication.instance()
        screens = list(application.screens()) if application else []
        if not screens:
            return None
        centre = QPoint(x + self.width() // 2, y + self.height() // 2)
        for screen in screens:
            if screen.geometry().contains(centre):
                return screen
        return min(screens, key=lambda s: _distance(s.geometry(), centre))


    # --- interaction --------------------------------------------------------

    def mousePressEvent(self, event) -> None:
        if event.button() == Qt.LeftButton:
            self._press = event.globalPosition().toPoint() - self.frameGeometry().topLeft()
            self._dragging = False
        event.accept()

    def mouseMoveEvent(self, event) -> None:
        if self._press is None:
            return
        target = event.globalPosition().toPoint() - self._press
        if not self._dragging:
            travelled = (target - self.pos()).manhattanLength()
            if travelled < CLICK_SLOP:
                return
            self._dragging = True
        self.place(target.x(), target.y())
        event.accept()

    def mouseReleaseEvent(self, event) -> None:
        if self._press is None:
            return
        was_dragging = self._dragging
        self._press = None
        self._dragging = False
        if was_dragging:
            self._on_moved(self.x(), self.y())
        else:
            self._on_activate()
        event.accept()


def _distance(rect, point: QPoint) -> int:
    """Squared distance from a point to a rectangle, zero inside it."""
    dx = max(rect.left() - point.x(), 0, point.x() - rect.right())
    dy = max(rect.top() - point.y(), 0, point.y() - rect.bottom())
    return dx * dx + dy * dy
